ParProcCo/job_scheduler.py

```diff
@@ -78,338 +80,505 @@ class SLURMSTATE(Enum):
-
-    def fetch_and_update_state(self, job_id: int) -> Optional[SLURMSTATE]:
-        job_info = self.client.get_job(job_id)
+        self.cluster_output_dir: Path | None = (
+            Path(cluster_output_dir) if cluster_output_dir else None
+        )
+        self.wait_timeout = wait_timeout
+        self.terminate_after_wait = terminate_after_wait
+
+    def fetch_and_update_state(
+        self, job_scheduling_info: JobSchedulingInformation
+    ) -> SLURMSTATE | None:
+        job_info = self.client.get_job(job_scheduling_info.job_id)
         job_id = job_info.job_id
         if job_id is None:
             raise ValueError(f"Job info has no job id: {job_info}")
         state = job_info.job_state
         slurm_state = SLURMSTATE[state] if state else None
-        try:
-            tres_alloc_str = job_info.tres_alloc_str
-            slots = (
-                int((tres_alloc_str.split(",")[1]).split("=")[1])
-                if tres_alloc_str
-                else None
-            )
 
-        except Exception:
-            logging.warning(
-                f"Failed to get slots for job {job_id}; setting slots to 0. Job info: {job_info}"
-            )
-            slots = None
+        tres_alloc_str = job_info.tres_alloc_str
+        if not tres_alloc_str:
+            cpus = None
+            gpus = None
+        else:
+            try:
+                cpu_match = re.search(r"cpu=(\d+)", tres_alloc_str)
+                cpus = int(cpu_match.group(1)) if cpu_match else None
+            except Exception as e:
+                print(e)
+                logging.warning(
+                    f"Failed to get cpus for job {job_id};"
+                    f" setting cpus to 0. Job info: {job_info}",
+                )
+                cpus = None
+            try:
+                gpu_match = re.search(r"gpu=(\d+)", tres_alloc_str)
+                gpus = int(gpu_match.group(1)) if gpu_match else None
+            except Exception as e:
+                print(e)
+                logging.warning(
+                    f"Failed to get gpus for job {job_id};"
+                    f" setting gpus to 0. Job info: {job_info}",
+                )
+                gpus = None
 
-        dispatch_time = job_info.start_time
+        start_time = job_info.start_time
         submit_time = job_info.submit_time
         end_time = job_info.end_time
-        if dispatch_time and submit_time and end_time:
-            time_to_dispatch = dispatch_time - submit_time
-            wall_time = end_time - dispatch_time
+
+        if start_time and submit_time and end_time:
+            time_to_dispatch = timedelta(seconds=start_time - submit_time)
+            wall_time = timedelta(seconds=end_time - start_time)
         else:
             time_to_dispatch = None
             wall_time = None
-        status_info = self.status_infos[job_id]
-        status_info.slots = slots
+
+        status_info = job_scheduling_info.status_info
+        assert status_info
```

Comment:
This wouldn't report for any further jobs if this fails - is that what is wanted? I think I'd go for something like:
```
if status_info is None:
    logging.error("No status info available for job %s", job_id)
    continue
```
And I'd have it after the line `logging.debug(f"Retrieving info for job {job_id}")`

Is there a reason assert is preferred here?

Reply:
This is to keep mypy happy. This `status_info` attribute is set by `_submit_jobs`.